src/MyHealth/symptoms.ts

```diff
@@ -22,3 +45,81 @@ export const determineHealthAssessment = (
   }
   return HealthAssessment.NotAtRisk
 }
+
+type WithADate = { date: Posix }
+const compareDatesAscending = (
+  { date: dateLeft }: WithADate,
+  { date: dateRight }: WithADate,
+) => {
+  return dateLeft < dateRight ? -1 : 1
+}
+const compareDatesDescending = (
+  { date: dateLeft }: WithADate,
+  { date: dateRight }: WithADate,
+) => {
+  return dateLeft < dateRight ? 1 : -1
+}
+
+const groupLogEntriesDaily = (allEntries: SymptomLogEntry[]): LogDataByDay => {
+  const groupedEntries: LogDataByDay = {}
+
+  allEntries.forEach((entry) => {
+    const { date } = entry
+    const entryDateBeginningOfDay = beginningOfDay(date)
+
+    if (groupedEntries[entryDateBeginningOfDay]) {
+      const newLogEntries = groupedEntries[entryDateBeginningOfDay].logEntries
+      newLogEntries.push(entry)
+      newLogEntries.sort(compareDatesAscending)
+      groupedEntries[entryDateBeginningOfDay].logEntries = newLogEntries
+    } else {
+      groupedEntries[entryDateBeginningOfDay] = {
+        logEntries: [entry],
+        checkIn: null,
+      }
+    }
+  })
+
+  return groupedEntries
+}
+
+const joinLogDataAndDailyCheckIns = (
+  logData: LogDataByDay,
+  dailyCheckIns: DailyCheckIn[],
+): LogDataByDay => {
+  dailyCheckIns.forEach((checkIn) => {
+    const { date } = checkIn
+    const checkInBeginningOfDay = beginningOfDay(date)
+
+    if (logData[checkInBeginningOfDay]) {
+      logData[checkInBeginningOfDay].checkIn = checkIn
+    } else {
+      logData[checkInBeginningOfDay] = {
+        logEntries: [],
+        checkIn,
+      }
+    }
+  })
+  return logData
+}
+
+export const serializeDailyLogData = (
+  symptomLogEntries: SymptomLogEntry[],
+  dailyCheckIns: DailyCheckIn[],
+): DayLogData[] => {
+  const symptomEntriesLogData = groupLogEntriesDaily(symptomLogEntries)
+  const dailyLogData = joinLogDataAndDailyCheckIns(
+    symptomEntriesLogData,
+    dailyCheckIns,
+  )
+
+  return Object.keys(dailyLogData)
+    .map((date: string) => {
+      const posixDate = Number(date)
+      return {
+        date: posixDate,
+        ...dailyLogData[posixDate],
+      }
+    })
+    .sort(compareDatesDescending)
+}
```

Comment:
This is intense 😮  🔥

Reply:
Sorry, I think I walked ourselves into a corner trying to simplify the data model we store so we can keep it separated as the feature asks of the data. Maybe we can re-shift how the two data models are stored and this can become a thing of the past?

Reply:
I can't think of a better way to do it 🤷

Unless maybe you could keep the data structures separate and then handle the "serialization" at the view layer. We could map over an array of the Posix times for the beginnings of the last 14 days and check if we have a daily check in or symptom log for that day and display it…

Reply:
Plus one to combining the `CheckInLogs` and the `SymptomLogs` in the view layer, this seems like a view specific concern.

Reply:
I agree that we could make the serialization closer to the view but in this case then we would need a context that holds logic for:
- updating log entries
- adding log entries

another one for:
- creating a check in
- creating a log entry

and have them both consume a context with all the log entries and check ins. Could this be second round improvement while we complete the functionality?